src/edge/simulator.ts

```diff
@@ -91,85 +56,90 @@ export class Simulator {
-                    await Executor.executeCMD(outputChannel, "pip", {shell: true}, "install --upgrade iotedgehubdev");
-                } catch (error) {
-                    if (outputChannel) {
-                        outputChannel.appendLine(`${Constants.failedInstallSimulator} ${error.message}`);
-                    }
-                    ret =  InstallReturn.Failed;
-                }
-            } else {
-                ret = InstallReturn.Canceled;
+    constructor(private context: vscode.ExtensionContext) {
+        if (Simulator.currentPlatform === "win32") {
+            const installedVersion: string = this.context.globalState.get(Simulator.simulatorVersionKey);
+            if (installedVersion) {
+                this.standaloneSimulatorExecutablePath = path.join(Simulator.WindowsStandaloneSimulatorFolder, installedVersion , Simulator.simulatorExecutableName);
             }
         } else {
-            ret = InstallReturn.NoPip;
+            this.standaloneSimulatorExecutablePath = Simulator.simulatorExecutableName;
         }
-        return ret;
     }
 
-    private static async simulatorInstalled(): Promise<boolean> {
-        return await Simulator.checkCmdExist("iotedgehubdev");
-    }
+    public async validateSimulatorUpdated(outputChannel: vscode.OutputChannel = null): Promise<void> {
+        let message: string;
+        let type: string = "";
+        const telemetryName = "simulatorUpdated";
 
-    private static async checkPipInstalled(): Promise<boolean> {
-        return await Simulator.checkCmdExist("pip");
-    }
+        const simulatorType = await this.simulatorInstalled();
+        if (simulatorType === SimulatorType.NotInstalled) {
+            message =  Constants.needSimulatorInstalledMsg;
+            type = "install";
+        } else {
+            const version: string | null = await this.getCurrentSimulatorVersion();
+            if (version && semver.valid(version)) {
+                const latestVersion: string | undefined = await this.getLatestSimulatorVersion();
+                if (latestVersion && semver.gt(latestVersion, version)) {
+                    message = `${Constants.updateSimulatorMsg} (${version} to ${latestVersion})`;
+                } else {
+                    return;
+                }
+            } else {
+                message = Constants.updateSimulatorMsg;
+            }
 
-    private static async checkCmdExist(cmd: string): Promise<boolean> {
-        try {
-            await Executor.executeCMD(null, cmd, { shell: true }, "--version");
-            return true;
-        } catch (error) {
-            return false;
+            if (simulatorType === SimulatorType.Pip) {
+                type = "upgradePipPackage";
+            } else {
+                type = "upgradeStandalone";
+            }
         }
-    }
 
-    constructor(private context: vscode.ExtensionContext) {
+        TelemetryClient.sendEvent(`${telemetryName}.${type}`);
+        try {
```

Comment:
how about move try upper to the entry of the method?

Reply:
OK

Reply:
We can use output channel to show unexpected errors